server/src/internalClusterTest/java/org/elasticsearch/snapshots/SnapshotShutdownIT.java

```diff
@@ -0,0 +1,530 @@
+            (handler, request, channel, task) -> masterTransportService.getThreadPool().generic().execute(() -> {
+                safeAwait(updateSnapshotStatusBarrier);
+                safeAwait(updateSnapshotStatusBarrier);
+                try {
+                    handler.messageReceived(request, channel, task);
+                } catch (Exception e) {
+                    fail(e);
+                }
+            })
+        );
+
+        final var clusterService = internalCluster().getCurrentMasterNodeInstance(ClusterService.class);
+        putShutdownForRemovalMetadata(primaryNode, clusterService);
+        unblockAllDataNodes(repoName); // lets the shard snapshot abort, but allocation filtering stops it from moving
+        safeAwait(updateSnapshotStatusBarrier); // wait for data node to notify master that the shard snapshot is paused
+
+        // abort snapshot (and wait for the abort to land in the cluster state)
+        final var deleteStartedListener = ClusterServiceUtils.addTemporaryStateListener(clusterService, state -> {
+            if (SnapshotDeletionsInProgress.get(state).getEntries().isEmpty()) {
+                return false;
+            }
+
+            assertEquals(SnapshotsInProgress.State.ABORTED, SnapshotsInProgress.get(state).forRepo(repoName).get(0).state());
+            return true;
+        });
+
+        final var deleteSnapshotFuture = startDeleteSnapshot(repoName, snapshotName); // abort the snapshot
+        safeAwait(deleteStartedListener);
+
+        safeAwait(updateSnapshotStatusBarrier); // process pause notification now that the snapshot is ABORTED
+
+        assertEquals(SnapshotState.FAILED, snapshotFuture.get(10, TimeUnit.SECONDS).getSnapshotInfo().state());
+        assertTrue(deleteSnapshotFuture.get(10, TimeUnit.SECONDS).isAcknowledged());
+
+        clearShutdownMetadata(clusterService);
+    }
+
+    public void testShutdownWhileSuccessInFlight() throws Exception {
+        final var primaryNode = internalCluster().startDataOnlyNode();
+        final var indexName = randomIdentifier();
+        createIndexWithContent(indexName, indexSettings(1, 0).put(REQUIRE_NODE_NAME_SETTING, primaryNode).build());
+
+        final var repoName = randomIdentifier();
+        createRepository(repoName, "mock");
+
+        final var clusterService = internalCluster().getCurrentMasterNodeInstance(ClusterService.class);
+        final var masterTransportService = MockTransportService.getInstance(internalCluster().getMasterName());
+        masterTransportService.addRequestHandlingBehavior(
+            SnapshotsService.UPDATE_SNAPSHOT_STATUS_ACTION_NAME,
+            (handler, request, channel, task) -> SubscribableListener
+
+                .<Void>newForked(
+                    l -> putShutdownMetadata(
+                        clusterService,
+                        SingleNodeShutdownMetadata.builder()
+                            .setType(SingleNodeShutdownMetadata.Type.REMOVE)
```

Comment:
A little, but `REPLACE` is tricky since you have to specify the replacement node, but that's mostly not relevant here. I left it with `REMOVE` and `SIGTERM` in these tests.